src/config/integrator-config.c

```diff
@@ -158,40 +186,112 @@ int Read_Integrator(XML_NODE node, void *config, __attribute__((unused)) void *c
+                    merror(REGEX_COMPILE, node[i]->content,
                        integrator_config[s]->location->error);
+                } else {
+                    snprintf(message, OS_FLSIZE,
+                        "Syntax error on regex: '%s': %d.",
+                        node[i]->content, integrator_config[s]->location->error);
+                    wm_strcat(output, message, '\n');
+                }
                 return(-1);
             }
         }
         else if(strcmp(node[i]->element, xml_integrator_group) == 0)
         {
+            if (integrator_config[s]->group) {
+                free(integrator_config[s]->group);
+            }
             os_strdup(node[i]->content, integrator_config[s]->group);
         } else if (strcmp(node[i]->element, xml_integrator_max_log) == 0) {
             if (!OS_StrIsNum(node[i]->content)) {
-                merror(XML_VALUEERR,node[i]->element, node[i]->content);
+                if (output == NULL) {
+                    merror(XML_VALUEERR, node[i]->element, node[i]->content);
+                } else {
+                    snprintf(message, OS_FLSIZE,
+                        "Invalid value for element '%s': %s.",
+                        node[i]->element, node[i]->content);
+                    wm_strcat(output, message, '\n');
+                }
                 return(OS_INVALID);
             }
 
             integrator_config[s]->max_log = atoi(node[i]->content);
 
             if (integrator_config[s]->max_log < 165 || integrator_config[s]->max_log > 1024) {
-                merror(XML_VALUEERR,node[i]->element, node[i]->content);
+                if (output == NULL) {
+                    merror(XML_VALUEERR, node[i]->element, node[i]->content);
+                } else {
+                    snprintf(message, OS_FLSIZE,
+                        "Invalid value for element '%s': %s.",
+                        node[i]->element, node[i]->content);
+                    wm_strcat(output, message, '\n');
+                }
                 return(OS_INVALID);
             }
-        } else
+        } else if (output == NULL)
         {
             merror(XML_INVELEM, node[i]->element);
             return(OS_INVALID);
         }
+        else {
+            snprintf(message, OS_FLSIZE,
+                "Invalid element in the configuration: '%s'.",
+                node[i]->element);
+            wm_strcat(output, message, '\n');
+            return(OS_INVALID);
+        }
         i++;
     }
 
     /* We must have at least one entry set */
     if(!integrator_config[s]->name)
     {
-        merror(XML_INV_INTEGRATOR);
+        if (output == NULL){
+            merror(XML_INV_INTEGRATOR);
+        } else {
+            wm_strcat(output, "Invalid integratord configuration.", '\n');
+        }
         return(OS_INVALID);
     }
 
     return(0);
 }
+
+int Test_Integratord(const char *path, char **output) {
+    IntegratorConfig **test_integrator = NULL;
```

Comment:
Should not test_integrator be an IntegratorConfig* instead of an IntegratorConfig**?

Reply:
It should be an integratorConfig**, because if a user introduces more than one integrator module in ossec.conf, the program will save the data in a vector. So for free the pointer later will need that be of type Integrator Config**.